Wallet settings, change password and download mnemonic

drillprop opened the pull request:
Fix #4243 #4319

drillprop committed:
wip

drillprop committed:
settings view ground work

drillprop committed:
uncomment route

drillprop committed:
add borderColor prop to banner

drillprop committed:
fix issues with modal paddings

drillprop committed:
create password form component, sign up modal refactoring

drillprop committed:
add export seed dialog and change password dialog

drillprop committed:
handle copy button

drillprop committed:
Merge remote-tracking branch 'upstream/accounts' into 4243-wallet-settings

drillprop committed:
show different settings for wallet users

drillprop committed:
refactor password dialogs

attemka commented on packages/atlas/src/components/_auth/PasswordCriterias/PasswordCriterias.tsx:
```diff
@@ -0,0 +1,122 @@
+      }
+
+      setPasswordRequirementsErrors(() => ({
+        length: getValidationState(password.length <= 9),
```
don't forget about max length (64)

attemka commented on packages/atlas/src/views/viewer/MembershipSettingsView/ChangePasswordDialog.tsx:
```diff
@@ -0,0 +1,222 @@
+const commonPasswordValidation = z
+  .string()
+  .regex(/^(?=.*[0-9])(?=.*[A-Z])(?=.*[!@#$%^&*()_+]).*$/, { message: 'Password has to meet requirements.' })
+  .min(9, { message: 'Password has to meet requirements.' })
```
max length

attemka commented on packages/atlas/src/views/viewer/MembershipSettingsView/ChangePasswordDialog.tsx:
```diff
@@ -0,0 +1,222 @@
+          />
+        </FormField>
+        <FormField label="Repeat Password" error={form.formState.errors.confirmPassword?.message}>
+          <Input placeholder="Repeat password" {...getInputProps('confirmPassword')} autoComplete="off" />
```
currently, this field is validated separately from the password field, but the only thing we need to validate for repeat password is that it matches the password in the field above.

attemka commented on packages/atlas/src/views/viewer/MembershipSettingsView/MembershipWallet/MembershipWallet.tsx:
```diff
@@ -0,0 +1,209 @@
+      <EntitySettingTemplate
+        isFirst
+        title="Membership address"
+        description="When your public membership was created, it was linked to a new substrate account address built on polkadot protocol.  This account holds all assets like tokens and NFTs that your membership accumulates. "
```
```suggestion
        description="When your public membership was created, it was linked to a new substrate account address built on polkadot protocol. This account holds all assets like tokens and NFTs that your membership accumulates. "
```

attemka reviewed (changes requested):
looks very nice. Only minor things to change

drillprop committed:
cr fixes

drillprop committed:
Update packages/atlas/src/views/viewer/MembershipSettingsView/MembershipWallet/MembershipWallet.tsx

Co-authored-by: attemka <attemka@gmail.com>

drillprop committed:
auth helpers further refactor

WRadoslaw commented on packages/atlas/src/views/viewer/MembershipSettingsView/ChangePasswordDialog.tsx:
```diff
@@ -0,0 +1,202 @@
+  encodedSeed: string
+  newPassword: string
+}
+export const changePassword = async ({
```
This can be contained inside auth provider helpers

WRadoslaw commented on packages/atlas/src/views/viewer/MembershipSettingsView/ChangePasswordDialog.tsx:
```diff
@@ -0,0 +1,202 @@
+    resolver: zodResolver(passwordAndRepeatPasswordSchema),
+  })
+
+  const [isFieldVisible, setIsFieldVisible] = useState<Record<PasswordInputNames, boolean>>({
```
This pattern started to be present in a few places already, maybe it will be better to transfer this logic to the input component?

WRadoslaw commented on packages/atlas/src/views/viewer/MembershipSettingsView/ExportSeedDialog.tsx:
```diff
@@ -0,0 +1,142 @@
+  link.download = 'mnemonic.json'
+  link.click()
+
+  URL.revokeObjectURL(url)
```
Shouldn't we also unmount the link, or will it be destroyed when fn completes?

WRadoslaw commented on packages/atlas/src/views/viewer/MembershipSettingsView/ExportSeedDialog.tsx:
```diff
@@ -0,0 +1,142 @@
+    shouldFocusError: true,
+  })
+  const { currentUser } = useAuth()
+  const [isPasswordVisible, setIsPasswordVisible] = useState(false)
```
Same as in ChangePassword modal

WRadoslaw commented on packages/atlas/src/views/viewer/MembershipSettingsView/ExportSeedDialog.tsx:
```diff
@@ -0,0 +1,142 @@
+    handleSubmit(async ({ password }) => {
+      try {
+        if (!currentUser) {
+          throw Error('Current user is not set')
```
In the ChangePassword dialog, we have the same check but do not throw an error

WRadoslaw reviewed (changes requested):
1. After changing my avatar I got redirected, but my membership was not refetched.
![image](https://github.com/Joystream/atlas/assets/92513933/221c6e71-e6ad-4ef9-ac57-a33ca53d3fd0)
The same thing applies for handle change, but since we use it in URL params after F5 is receive empty page 
![image](https://github.com/Joystream/atlas/assets/92513933/bcedb34e-db88-40a1-853c-81817c535af5)

2. We display tx fee even tho nothing was changed (public profile tab)
![image](https://github.com/Joystream/atlas/assets/92513933/ae66b66b-032d-4b44-bdee-7579400ac54f)

3. Why do we need action bar in wallet tab if every action performed in this tab is performed inside a specific dialog?

4. @dmtrjsg @attemka  I know that it's possible for someone determined to retrieve a mnemonic (with local storage) after the user forgot to log out or smth to do it. But having password auth for seed export and right under its action to change the password without knowing the previous one is pointless.

drillprop committed:
reset form on cancel

drillprop committed:
Merge remote-tracking branch 'upstream/accounts' into 4243-wallet-settings

drillprop committed:
cr fixes

drillprop commented:
> 1. After changing my avatar I got redirected, but my membership was not refetched.
>    ![image](https://user-images.githubusercontent.com/92513933/246066851-221c6e71-e6ad-4ef9-ac57-a33ca53d3fd0.png)
>    The same thing applies for handle change, but since we use it in URL params after F5 is receive empty page
>    ![image](https://user-images.githubusercontent.com/92513933/246067398-bcedb34e-db88-40a1-853c-81817c535af5.png)

fixed

> 2. We display tx fee even tho nothing was changed (public profile tab)
>    ![image](https://user-images.githubusercontent.com/92513933/246072284-ae66b66b-032d-4b44-bdee-7579400ac54f.png)

fixed 

> 3. Why do we need action bar in wallet tab if every action performed in this tab is performed inside a specific dialog?

I asked Kuba about this, and he said that we will keep the action bar, but the tooltip above the publish changes will just always say, that's is nothing to publish.

drillprop committed:
cr fixes 2

dmtrjsg commented:
@KubaMikolajczyk we just discussed this comment here and arrived to the need to request password old password for changing password. Could you work directly with @drillprop to add this in pls ..

KubaMikolajczyk commented:
@drillprop the password additional step in change password user story is added here ✅ https://www.figma.com/file/yhZpTHdf1sxJx13uRZ71GV/Membership-profile?type=design&node-id=2912-46498&t=CPxJwOpMmWqPNTUz-4

drillprop committed:
add additional step to ChangePassword

drillprop commented:
@WRadoslaw all issues should be fixed now, ready to review

dmtrjsg commented:
@KubaMikolajczyk 

1. We don't need publish/ cancel on the `wallet tab` as there are no inputs to change on the actual page? 
2. Do we want to add "back to profile"? As there is no way to go back now (cancel button is wired to do it but feels suboptimal)
3. The cancel button is only valid if you edit smth but shown always?

The argument is to show the bottom panel only on change of inputs.. wdyt?

drillprop committed:
action bar adjustments, add unsaved changes dialog

drillprop committed:
fix viewer layout

KubaMikolajczyk commented:
@dmtrjsg implemented solutions for your feedback above. Edited few components and added documentation for action bar behavior logic. Also all of the pages are now updated with the new logic. 
🎥 Loom update: https://www.loom.com/share/00925d975de8440cb3c1b649d9885764?sid=41b52e4a-ee96-45ae-acfd-0392018758fa
📄 Membership profile changelog: https://www.figma.com/file/yhZpTHdf1sxJx13uRZ71GV/Membership-profile?type=design&node-id=2912-46498&t=NL8Mje6NhP5iQSJu-4
📄 My channel profile changelog: https://www.figma.com/file/zi5siPUXme0Fivq2i40XBd/My-channel?type=design&node-id=703-5383&mode=design&t=ahFXFmRVHCGuSKja-4

drillprop commented:
Draft, I need to apply new action bar changes

drillprop committed:
action bar changes

drillprop committed:
Merge remote-tracking branch 'upstream/accounts' into 4243-wallet-settings

drillprop committed:
action bar adjustments

WRadoslaw commented on packages/atlas/src/views/viewer/MembershipSettingsView/ChangePasswordDialog.tsx:
```diff
@@ -0,0 +1,219 @@
+    })()
+  }, [currentUser, displaySnackbar, verifyPasswordForm])
+
+  const changePasswordform = useForm<PasswordStepForm>({
```
```suggestion
  const changePasswordForm = useForm<PasswordStepForm>({
```

WRadoslaw commented on packages/atlas/src/views/viewer/MembershipSettingsView/ChangePasswordDialog.tsx:
```diff
@@ -0,0 +1,219 @@
+  confirmPassword: string
+}
+
+export const getMnemonicFromeEmailAndPassword = async (email: string, password: string) => {
```
Don't we have anything similar in auth provder helpers?

drillprop replied:
We don't but I moved the function to helpers

WRadoslaw reviewed (changes requested):
1. If we change any field in public profile tab and reverse it the fee will not zero out
![image](https://github.com/Joystream/atlas/assets/92513933/ff8c5723-6b6b-4890-9227-94f1a739a65a)

2. I cannot change password
![image](https://github.com/Joystream/atlas/assets/92513933/efdc6d56-8f66-401d-8cfc-443b5050730f)

drillprop committed:
cr fixes

drillprop committed:
fix fee issue

drillprop committed:
change password fix?

drillprop commented:
> 1. If we change any field in public profile tab and reverse it the fee will not zero out
>    ![image](https://user-images.githubusercontent.com/92513933/249046014-ff8c5723-6b6b-4890-9227-94f1a739a65a.png)

Fixed

> 2. I cannot change the password
>    ![image](https://user-images.githubusercontent.com/92513933/249046363-efdc6d56-8f66-401d-8cfc-443b5050730f.png)

I wasn't able to reproduce it, but I used the same timestamp logic that we have in the login and signup functions, so hopefully, it will be fine now.
